PatternedMeshGenerator boundary stitching patch

pbehne opened the pull request:
This commit patches the bug described in #23312 which results in incorrect mesh stitching in the PatternedMeshGenerator when the mesh tiles provided in the 'inputs' parameter do not have the same boundary ids. A check is added to detect this issue, and if present, the boundary ids of all tile meshes are changed to be common, previously unused boundary ids. This allows for the tiles to be stitched together correctly. Additionally, the class was cleaned up to incorporate recommended code standards.

closes #23312

pbehne committed:
PatternedMeshGenerator boundary stiching patch

This commit patches the bug described in #23312 which results in incorrect mesh
stitching in the PatternedMeshGenerator when the mesh tiles provided in the
'inputs' parameter do not have the same boundary ids.
A check is added to detect this issue, and if present, the boundary ids of all
tile meshes are changed to be common, previously unused boundary ids.
This allows for the tiles to be stitched together correctly.
Additionally, the class was cleaned up to incorporate recommended code
standards.

closes #23312

pbehne commented:
@loganharbour

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1330080/) on 63266d1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23326/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1330094/) on 63266d1 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c677e7</th>
      <th colspan=4>#23326 63266d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.62%</td>
      <td>84.61%</td>
      <td>-0.02%</td>
      <td>52.38%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>81993</td>
      <td>82000</td>
      <td>+7</td>
      <td>22</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14898</td>
      <td>14918</td>
      <td>+20</td>
      <td>20</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23326/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23326/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23326/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 52.38% is less than the suggested 90.0%



This comment will be updated on new commits.

loganharbour reviewed (changes requested):
Lots of cleanup changes that I think are straight forward

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1331742/) on 882fb21 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23326/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23326/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9ef9524ac3cc5af54681e25fd1660b5a51ca5604`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1331742/) on 882fb21 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23326/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23326/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9ef9524ac3cc5af54681e25fd1660b5a51ca5604`

pbehne commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,36 +101,36 @@ PatternedMeshGenerator::generate()
+    if (top_id == -123)
+      mooseError("The top boundary has not been initialized properly.");
+    if (bottom_id == -123)
+      mooseError("The bottom boundary has not been initialized properly.");
```
Incorporated.

pbehne committed:
Combined variable into vectors, implemented better invalid boundary id checking.

pbehne commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,35 +104,36 @@ PatternedMeshGenerator::generate()
+    for (const auto side : make_range(4))
+    {
+      if (ids[side] == Moose::INVALID_BOUNDARY_ID)
+        paramError("inputs",
+                   "The '",
+                   boundary_param_names[side],
+                   "' parameter with value '",
+                   boundary_names[side],
+                   "' does not exist in input mesh '",
+                   _input_names[i],
+                   "'");
```
@loganharbour, let me know if you want any additional changes to this one. It's a bit tricky because an error results from a combination of TWO input parameters (i..e, when one of left_boundary, right_boundary, top_boundary, or bottom_boundary does not match up with a boundary name of any of the meshes in inputs).

pbehne replied:
See comment above about generate_unknown parameter in getBoundaryIDs function.

loganharbour replied:
Talk to me about this in person, I'm a bit confused at what you're trying to say in the first comment. Also - we might want to check to make sure that all of the options in [left, right, top, bottom] are unique.

pbehne commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,35 +104,36 @@ PatternedMeshGenerator::generate()
 
-  std::vector<boundary_id_type> ids =
-      MooseMeshUtils::getBoundaryIDs(*_meshes[0], boundary_names, true);
+    const auto ids = MooseMeshUtils::getBoundaryIDs(*_meshes[i], boundary_names, false);
```
In order to properly trigger the paramError for "inputs", I changed the "generate_unknown" parameter in getBoundaryIDs from true to false. When true, a new boundary_id is generated if no match to the requested boundary name is found. When false, invalid_id is returned, which is what triggers the paramError below.

By changing generate_unknown to False, so of the tests in the reactor module are failing because the paramError is being triggered. My hypothesis is that these tests rely on new boundary ids being generated in a way that we are not taking into account.

loganharbour replied:
FYI @miaoyinb and @yjung-anl, looks like we've got a pattern stitch that relies on some undefined behavior that we cleaned up (see https://civet.inl.gov/job/1331932/)

pbehne replied:
The failing tests have been updated to work correctly.

loganharbour commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -128,6 +145,26 @@ PatternedMeshGenerator::generate()
+          break;
+      }
+
+  for (const auto i : index_range(_meshes))
+    for (const auto side : make_range(4))
+      MeshTools::Modification::change_boundary_id(
+          *_meshes[i], input_bids[i][side], stitch_bids[side]);
```
```suggestion
  if (!have_common_ids)
    for (const auto i : index_range(_meshes))
      for (const auto side : make_range(4))
        MeshTools::Modification::change_boundary_id(
            *_meshes[i], input_bids[i][side], stitch_bids[side]);
```

This makes sense right?

pbehne replied:
Yup, will incorporate.

loganharbour reviewed (changes requested):
We'll chat in a bit. I think the last thing we need is to fix the failures and add some error testing

pbehne committed:
Added check for unique boundary names, moved final boundary renaming into not have_common_ids conditional.

pbehne committed:
Added note to documentation describing changes.

pbehne committed:
Added new tests for PatternedMeshGenerator, updated existing ones.

loganharbour commented:
@miaoyinb or @yjung-anl can you please verify the changes in a0c894eb01286b052c043841b358436b7250e32c

miaoyinb commented:
> @miaoyinb or @yjung-anl can you please verify the changes in [a0c894e](https://github.com/idaholab/moose/commit/a0c894eb01286b052c043841b358436b7250e32c)

I think this is for @yjung-anl.

yjung-anl commented:
@pbehne 
The changes look good to me!
Thank you for identifying and fixing this issue!

pbehne commented on modules/reactor/test/tests/meshgenerators/assembly_mesh_generator/assembly_only.i:
```diff
@@ -20,17 +20,37 @@
-    region_ids='1 2 5; 11 12 15'
+    region_ids = '1 2 5; 11 12 15'
+  []
+  [pin1]
```
@yjung-anl
I am trying to fix all the remaining tests in the reactor module that are failing. See https://civet.inl.gov/job/1338882/ for failure details. With the new error checks in PatternedMeshGenerator, we need to make sure that any MeshGenerators in AssemblyMeshGenerator/inputs have boundary sidesets named 'left', 'right', 'top', and 'bottom'. To do this, I renamed pin1 to pin1_a and ran it through a SideSetsFromNormalsGenerator to get the desired boundaries. However, the AssemblyMeshGenerator does not like this and gives the following error:

```
The reactor_params input must be a ReactorMeshParams type MeshGenerator
Please check that a valid definition and name of ReactorMeshParams has been provided.
```

Could you please look into a fix for this?

miaoyinb replied:
This is because the mesh metadata generated by `PinMeshGenerator` are lost when you modify the mesh using SideSetsFromNormalsGenerator. @loganharbour @GiudGiud  and I are working on the mesh metadata forwarding capability development that will eventually solve this issue. I will help fix this once #23353 is merged.

pbehne replied:
We were able to implement a fix in the source code that does not requires modification of the test input files.

pbehne committed:
Added support for non-unique boundaries.

The newly added 'require_unique_boundaries' parameter controls whether the
left/right/top/bottom boundary name parameter values are requires to be unique
from each other.
The parameter is true by default, but is set to false and then suppressed in
the derived CartesionIDPatternedMeshGenerator class.

pbehne commented:
Note that a Griffin test in the controlled app tests fails. PR 1055 in the Griffin repo corrects this issue.

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -128,6 +158,29 @@ PatternedMeshGenerator::generate()
+  else
+  {
+    for (boundary_id_type id = 0; id != Moose::INVALID_BOUNDARY_ID; ++id)
+      if (!used_boundary_ids.count(id))
+      {
+        stitch_bids.push_back(id);
+        if (stitch_bids.size() == 4)
+          break;
+      }
```
I think there's a util to get the next free boundary id. and if there is not we should add one

loganharbour replied:
We use it in a few places but it's used in different ways, for example in RenameBoundaryGenerator we have to also check again an extra set of bids

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,35 +117,36 @@ PatternedMeshGenerator::generate()
-    mooseError("The top boundary has not been initialized properly.");
-  if (bottom == -123)
-    mooseError("The bottom boundary has not been initialized properly.");
+    for (const auto j : index_range(boundary_names))
```
add comments

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -128,6 +158,29 @@ PatternedMeshGenerator::generate()
+    for (const auto i : index_range(_meshes))
+      for (const auto side : make_range(4))
+        MeshTools::Modification::change_boundary_id(
+            *_meshes[i], input_bids[i][side], stitch_bids[side]);
```
this could be problematic if input_bids[i] is the same on all sides.

Then you end up with a no sideset with stitch_bids[0, 1 or 2]

loganharbour replied:
Not possible by default: https://github.com/idaholab/moose/blob/470e4353ecbd7163dd55ed236bcd81eb3cc485ba/framework/src/meshgenerators/PatternedMeshGenerator.C#L91

We even discussed this (this error is what caused problems with reactor)

why must you hijack my PR

GiudGiud replied:
no I think it breaks as I say
```
[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 5
    ny = 5
    subdomain_ids = '1 1 1 1 1
                     1 1 1 1 1
                     1 1 1 1 1
                     1 1 1 1 1
                     1 1 1 1 1
                     '
  []

  [gmg2]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 5
    ny = 5
    subdomain_ids = '2 2 2 2 2
                     2 2 2 2 2
                     2 2 2 2 2
                     2 2 2 2 2
                     2 2 2 2 2
                     '
    # The following triggers generation of new common boundary ids in
    # PatternedMeshGenerator
    boundary_id_offset = 1
  []

  [2_sides_same]
    type = RenameBoundaryGenerator
    input = gmg2
    old_boundary = 'left top'
    new_boundary = 'left left'
  []
  [2_sides_same2]
    type = RenameBoundaryGenerator
    input = gmg
    old_boundary = 'left top'
    new_boundary = 'left left'
  []

  [pmg]
    type = PatternedMeshGenerator
    inputs = '2_sides_same2 2_sides_same'
    pattern = '1 0 ;
               0 1'
    top_boundary = 'left'
    require_unique_boundaries = false
  []
[]
```

```
In UnstructuredMesh::stitch_meshes:
This mesh has 11 nodes on boundary 8.
Other mesh has 0 nodes on boundary 7.
Minimum edge length on both surfaces is 0.2.
In UnstructuredMesh::stitch_meshes:
Found 0 matching nodes.
```

pbehne replied:
Setting 'require_unique_boundaries' to true (the default value) fixes this issue ;)

GiudGiud replied:
unless you forbid users from setting it to false then there's an issue.

It fails silently too which is more problematic

pbehne commented on framework/doc/content/source/meshgenerators/PatternedMeshGenerator.md:
```diff
@@ -7,6 +7,10 @@
 adds the ability to create a tile pattern from an arbitrary number of input meshes.
 
+!alert note
+The MeshGenerators provided in inputs are no longer required to have the same boundary ids for a given boundary.
```
Correct. I will update for clarity.

loganharbour commented on framework/doc/content/source/meshgenerators/PatternedMeshGenerator.md:
```diff
@@ -7,6 +7,10 @@
 
+!alert note
+The MeshGenerators provided in inputs are no longer required to have the same boundary ids for a given boundary.
+Boundary ids for all input MeshGenerators are temporarily renamed to a common unused set of ids. The resulting PatternedMeshGenerator has boundary ids that are the same as the first entry of the inputs parameter.
```
The inputs aren't modified, re: temporarily

pbehne commented on framework/doc/content/source/meshgenerators/PatternedMeshGenerator.md:
```diff
@@ -7,6 +7,10 @@
 
+!alert note
+The MeshGenerators provided in inputs are no longer required to have the same boundary ids for a given boundary name.
+In this case, boundary ids for all (cloned) input MeshGenerators are renamed to a common unused set of ids. The resulting PatternedMeshGenerator has boundary ids that are the same as the first entry of the passed inputs parameter.
```
I know what you are trying to say, but I do not think this wording is correct from all interpretations. Each of the meshes in the grid is technically a clone inside PatterendMeshGenerator::generate (see [here](https://github.com/idaholab/moose/blob/2509317e71abacfdea1b50465f28a7ca62170326/framework/src/meshgenerators/PatternedMeshGenerator.C#L120)). In the case that these meshes do not have common boundary ids, the boundary ids of these clones are renamed once to common, previously unused boundary ids, not to the boundary ids of the top left mesh (see [here](https://github.com/idaholab/moose/blob/2509317e71abacfdea1b50465f28a7ca62170326/framework/src/meshgenerators/PatternedMeshGenerator.C#L162-L182)).  It is only the stitched-together mesh (that is returned) that has the boundary ids of the original, top left mesh (see [here)](https://github.com/idaholab/moose/blob/2509317e71abacfdea1b50465f28a7ca62170326/framework/src/meshgenerators/PatternedMeshGenerator.C#L245-L251). If you still do not like my use of "The resulting PatternedMeshGenerator has boundary ids..." above, I think "The resulting stitched mesh has boundary ids" would work.

pbehne replied:
Also, your use of "top left" is fine.

pbehne commented on framework/doc/content/source/meshgenerators/PatternedMeshGenerator.md:
```diff
@@ -7,6 +7,10 @@
 
+!alert note
+The MeshGenerators provided in inputs are no longer required to have the same boundary ids for a given boundary name.
+In this case, boundary ids for all (cloned) input MeshGenerators are renamed to a common unused set of ids. The resulting PatternedMeshGenerator has boundary ids that are the same as the first entry of the passed inputs parameter.
```
```suggestion
In this case, boundary ids for all (cloned) input MeshGenerators are renamed to a common unused set of ids. The resulting stitched patterned mesh has boundary ids that are the same as the first entry of the passed inputs parameter (top left mesh in the grid).
```

pbehne commented:
@GiudGiud I had some nice debug print statements in PatternedMeshGenerator.C, but removed them to make CIVET happy. If it will help your review, feel free to check out 17c0dd7.

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,35 +116,74 @@ PatternedMeshGenerator::generate()
-  if (bottom == -123)
-    mooseError("The bottom boundary has not been initialized properly.");
+  // Error check on lengths of input_bids_unique
+  if (lengths_set.size() != 1)
```
you could check the content to be 1 or 4 just below

pbehne replied:
See our above conversation. I am leaving this check out for more user flexibility.

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -128,6 +195,43 @@ PatternedMeshGenerator::generate()
+    // Make all inputs have common boundary ids
+    for (const auto i : index_range(_meshes))
+      for (const auto side : index_range(stitch_bids))
+      {
```
```suggestion
```

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -128,6 +195,43 @@ PatternedMeshGenerator::generate()
+      {
+        MeshTools::Modification::change_boundary_id(
+            *_meshes[i], input_bids_unique[i][side], stitch_bids[side]);
+      }
```
```suggestion
```

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,35 +116,74 @@ PatternedMeshGenerator::generate()
+      // meshes results in undefined behaviour when stitching
+      if (i == 0)
+      {
+        if (!id_set.count(ids[side]))
```
you dont need id_set.
You're storing this information of "I have seen this before" in another vector (input_bids_unique) and a map seen_bid_to_index_map

pbehne replied:
How do you recommend I replace the
`if (!id_set.count(ids[side]))`
statements? Use std::find on input_bids_unique and query that result instead?

GiudGiud replied:
yeah there's a count on vectors and probably just find on the map

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,35 +116,74 @@ PatternedMeshGenerator::generate()
+    }
 
-  boundary_id_type left = ids[0], right = ids[1], top = ids[2], bottom = ids[3];
+    lengths_set.insert(input_bids_unique.size());
```
I wouldnt keep track of length_sets like this.
If you make it a single integer, you can error as soon as you hit the inconsistent size, and you can return the name of the problematic inputs

pbehne replied:
I will modify to use a single integer. However, I don't think there is a quick way to point to a single input mesh as the problematic input without adding additional checks. For example: if there are two input meshes with a different length of input_bids_unique, which one is considered the problematic input?

GiudGiud replied:
the first one you hit in the loop? Or the first one that is not 1 or 4

GiudGiud replied:
I guess it s why it s easier to check for 1/4 than to check for "all the same"

GiudGiud commented on test/tests/meshgenerators/patterned_mesh_generator/tests:
```diff
@@ -1,13 +1,36 @@
+      detail = 'the boundary name does not exist'
+      mesh_mode = 'REPLICATED'
+    []
+  []
```
add an error test for the inconsistent size of boundary_ids

pbehne replied:
I am actually having a difficult time coming up with an input that triggers this error without hitting the error (here)[https://github.com/idaholab/moose/blob/74817397957818d058304850d3fca6d7ceecd352/framework/src/meshgenerators/PatternedMeshGenerator.C#L136-L145] first. Do you have any ideas? If not, I think the inconsistent size test can be removed.

GiudGiud replied:
just use one with 4 sides and one with 1 side all around?

pbehne replied:
In such a case, there is no correct way to set left/right/top/bottom_boundary. See below:

`[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 5
    ny = 5
    subdomain_ids = '1 1 1 1 1
                     1 1 1 1 1
                     1 1 1 1 1
                     1 1 1 1 1
                     1 1 1 1 1
                     '
  []

  [gmg2]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 5
    ny = 5
    subdomain_ids = '2 2 2 2 2
                     2 2 2 2 2
                     2 2 2 2 2
                     2 2 2 2 2
                     2 2 2 2 2
                     '
  []

  [all_sides_same]
    type = RenameBoundaryGenerator
    input = 'gmg2'
    old_boundary = 'left right top bottom'
    new_boundary = 'left left left left'
  []

  [pmg]
    type = PatternedMeshGenerator
    inputs = 'gmg all_sides_same'
    pattern = '1 0 ;
               0 1'
  []
[]`

The error:

`*** ERROR ***
(Mesh/pmg/inputs):
    The 'right_boundary' parameter with value 'right' does not exist in input mesh 'all_sides_same'`

Now if I set left/right/top/bottom_boundary all to 'left':

`  [pmg]
    type = PatternedMeshGenerator
    inputs = 'gmg all_sides_same'
    pattern = '1 0 ;
               0 1'
    left_boundary = left
    right_boundary = left
    top_boundary = left
    bottom_boundary = left
  []`

This runs because the boundary id for the 'left' mesh in each input mesh is a single number, thus the error is not triggered.

pbehne commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -76,8 +76,36 @@ PatternedMeshGenerator::generate()
   // Reserve spaces for all the meshes
   _meshes.reserve(_input_names.size());
 
+  // Getting the boundaries provided by the user
```
Remember that even with that check in place, I still found a way to break the mesh generator without triggering the check. Because we don't have a way to catch these cases (without normals checking), I don't see a reason to leave the check in place since there are valid cases 1 > num_boundaries > 4 that produce correct meshes. Still want me to add the check back in?

pbehne replied:
Also, I did update the manual page for PatternedMeshGenerator to inform the user specifically what to look for when the mess up their inputs.

GiudGiud replied:
As you want. I think it s a fine limitation

pbehne replied:
Will leave the check out.

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -76,8 +76,36 @@ PatternedMeshGenerator::generate()
+  std::vector<std::vector<boundary_id_type>> input_bids_unique(_input_names.size());
+
+  // For an error check on input_bids_unique later
+  size_t set_length;
```
```suggestion
  size_t set_length = 0;
```

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,35 +116,70 @@ PatternedMeshGenerator::generate()
-  if (bottom == -123)
-    mooseError("The bottom boundary has not been initialized properly.");
+        if (std::count(input_bids_unique[i].begin(), input_bids_unique[i].end(), ids[side]) > 0)
+        {
```
```suggestion
```

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,35 +116,70 @@ PatternedMeshGenerator::generate()
+        if (std::count(input_bids_unique[i].begin(), input_bids_unique[i].end(), ids[side]) > 0)
+        {
+          input_bids_unique[i].push_back(ids[side]);
+        }
```
```suggestion
```

GiudGiud reviewed (commented):
good to go after this and the extra test

GiudGiud commented:
Once you're done with the seg fault and the other test, we should look at condensing this commit history a little (squashing)

pbehne commented:
@GiudGiud, hold off on merging until I give you the go ahead. I had to make a change that I shouldn't have, so I want to look through all changes one more time to make sure I haven't missed anything.

pbehne commented:
@GiudGiud, I am ready for merging if you are.

GiudGiud commented on framework/doc/content/source/meshgenerators/PatternedMeshGenerator.md:
```diff
@@ -7,6 +7,13 @@
 
+!alert note
+The MeshGenerators provided in inputs are no longer required to have the same boundary ids for a given boundary name.
+In this case, boundary ids for all (cloned) input MeshGenerators are renamed to a common unused set of ids. The resulting stitched patterned mesh has boundary ids that are the same as the first entry of the passed inputs parameter (top left mesh in the grid).
```
```suggestion
In this case, boundary ids for all (cloned) input MeshGenerators are first renamed to a common unused set of ids. Them the resulting stitched patterned mesh will be changed again to boundary ids that are the same as the first entry of the passed inputs parameter (top left mesh in the grid).
```
Last sentence does not work
See our input
```
    inputs = 'fmg fmg2'
    pattern = '1 0 0 0 0 0 0 0 0 0 0 0 0 0 ;
               0 1 1 0 0 0 0 0 0 0 0 1 1 0 ;
               0 1 1 1 0 0 0 0 0 0 1 1 1 0 ;
               0 1 0 1 1 0 0 0 0 1 1 0 1 0 ;
               0 1 0 0 1 1 0 0 1 1 0 0 1 0 ;
               0 1 0 0 0 1 1 1 1 0 0 0 1 0 ;
               0 1 0 0 0 0 1 1 0 0 0 0 1 0 ;
               0 1 0 0 0 0 0 0 0 0 0 0 1 0 ;
               0 1 0 0 0 0 0 0 0 0 0 0 1 0 ;
               0 1 0 0 0 0 0 0 0 0 0 0 1 0 ;
               0 1 0 0 0 0 0 0 0 0 0 0 1 0 ;
               0 0 0 0 0 0 0 0 0 0 0 0 0 0'
```

first input and top-left pattern could be not the same

GiudGiud commented on framework/doc/content/source/meshgenerators/PatternedMeshGenerator.md:
```diff
@@ -7,6 +7,13 @@
+In this case, boundary ids for all (cloned) input MeshGenerators are renamed to a common unused set of ids. The resulting stitched patterned mesh has boundary ids that are the same as the first entry of the passed inputs parameter (top left mesh in the grid).
+
+!alert note
+Although best efforts have been made to ensure that the PatternedMeshGenerator fails with a descriptive error message when provided with invalid or ambiguous input, there are some expections that will result in silent failure (the input meshes are stitched together incorrectly and the program continues to run). If this occurs, this issue is likely that one of the left/right/top/bottom_boundary in ...
```
```suggestion
Although best efforts have been made to ensure that the PatternedMeshGenerator fails with a descriptive error message when provided with invalid or ambiguous input, there are some expections that will result in silent failure (the input meshes are stitched together incorrectly and the program continues to run). If this occurs, this issue is likely that one of the left/right/top/bottom_boundary input parameters is set to an incorrect value (i.e., left_boundary = 'right', when the 'right' boundary does not have outward normal (-1, 0, 0)). Check for mistakes like these when troubleshooting errors. Additionally, it should be confirmed that the values of the left/right/top/bottom_boundary parameters are correct for EVERY mesh provided in the 'inputs' parameter (i.e., each mesh in 'inputs' should have the same boundary name, ids may differ, and locations).
```

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -76,8 +76,36 @@ PatternedMeshGenerator::generate()
+  // Using a vector of vectors instead of vector of sets to preserve insertion order
+  std::vector<std::vector<boundary_id_type>> input_bids_unique(_input_names.size());
+
+  // For an error check on input_bids_unique later
```
```suggestion
  // For an error check on the number of uniquely named boundaries to stitch
```

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -88,35 +116,68 @@ PatternedMeshGenerator::generate()
-                                              getParam<BoundaryName>("top_boundary"),
-                                              getParam<BoundaryName>("bottom_boundary")};
+    // Keep track of used IDs so we can later find IDs that are unused across all meshes
+    used_boundary_ids.insert(ids.begin(), ids.end());
```
this isnt a full list of used ids. It's only ids used for the the names op top/left/...

In ids, you need to get ALL ids in those meshes to find ones that are never used. 

Otherwise when you change ids you ll hit existing ids that are not part of top/left/...

GiudGiud replied:
but then your use of ids below is different. 

I reckon you should create a new all_ids set here, then use all_ids for used_boundary_ids

pbehne commented:
@GiudGiud, please check changes in latest commit. If all looks good, I will squash into the previous commit.

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -76,8 +76,36 @@ PatternedMeshGenerator::generate()
+      _input_names.size(), std::vector<boundary_id_type>(4, Moose::INVALID_BOUNDARY_ID));
+  bool have_common_ids = true;
+
+  // Keep track of used boundary ids to generate new, unused ones later (if needed)
+  std::set<boundary_id_type> all_boundary_ids;
```
move this down too

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -76,8 +76,36 @@ PatternedMeshGenerator::generate()
+
+  // Given a boundary name (i.e., 'left_boundary'), this will give the correct
+  // index to get the correct boundary id to later use for boundary stitching
+  std::map<std::string, size_t> boundary_name_to_index_map;
```
move it down next to where it s filled too

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -76,8 +76,36 @@ PatternedMeshGenerator::generate()
+                                                    getParam<BoundaryName>("bottom_boundary")};
+  const std::vector<std::string> boundary_param_names = {
+      "left_boundary", "right_boundary", "top_boundary", "bottom_boundary"};
+  const std::set<std::string> boundary_names_set(boundary_names.begin(), boundary_names.end());
```
ugh was left as a pending 2 hours ago

pbehne replied:
Not anymore. I have removed it.

pbehne commented:
@loganharbour, @GiudGiud, Any idea what's going on with the Griffin failures? I don't see how they are related to my PR.

GiudGiud commented:
Unrelated

GiudGiud commented:
Logan is assigned so he'll merge your PR when appropriate. You can move on to something else at this point

loganharbour commented:
Will go in today when it's all green.
